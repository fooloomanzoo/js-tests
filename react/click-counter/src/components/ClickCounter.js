import React from 'react';

/**
 * css import
 */
import style from '../style/click-counter.css';

const STYLECLASS = {
  ACTIVE: 'active',
  BASIC: 'basic'
};

/**
 * component declaration
 */
export default class ClickCounter extends React.Component {

  constructor(props) {
    super(props);
    // bind handler to `this`
    this.incrementCounter = this.incrementCounter.bind(this);
    this.onDownHandler = this.onDownHandler.bind(this);
    this.keyDownHandler = this.keyDownHandler.bind(this);
    this.removeActiveState = this.removeActiveState.bind(this);
    // initiate `state`
    this.state = {
      count: 0,
      active: false,
      class: style[STYLECLASS.BASIC]
    };
  }

  /**
   * handle key down events
   * @argument e Causing event
   */
  keyDownHandler(e) {
    if (e && e.keyCode === 13) {
      this.incrementCounter();
    }
  }

  /**
   * handle mouse-down- and touch-start-events
   * @argument e Causing event
   */
  onDownHandler(e) {
    if (e && e.keyCode === 13) {
      this.incrementCounter();
    }
  }

  /**
   * increment counter
   * @argument e Causing event
   */
  incrementCounter(e) {
    this.setState({
      count: this.state.count + 1,
      active: true,
      class: style[STYLECLASS.BASIC] + ' ' + style[STYLECLASS.ACTIVE]
    });
  }

  /**
   * removes active state from the element
   */
  removeActiveState() {
    this.setState({
      active: false,
      class: style[STYLECLASS.BASIC]
    });
  }

  /**
   * render component
   */
  render() {
    return (
      <h1 className= { this.state.class }
        onMouseDown = { this.incrementCounter }
        onTouchStart = { this.incrementCounter }
        onKeyDown = { this.keyDownHandler }
        onMouseUp = { this.removeActiveState }
        onTouchEnd = { this.removeActiveState }
        onKeyUp = { this.removeActiveState }
        aria-label = 'counter'
        aria-live = 'polite'
        aria-relevant = 'text'
        tabIndex = { 0 }>
        Clicks: { this.state.count }
      </h1>
    );
  }
}
