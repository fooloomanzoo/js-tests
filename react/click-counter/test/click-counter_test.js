import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import ClickCounter from '../src/components/ClickCounter.js';

describe('Basic Functionality', () => {
  it('Renders ClickCounter component correctly', () => {
    const tree = renderer
      .create(<ClickCounter />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Clicking increases `state.count` in ClickCounter and sets `state.active`', () => {
    const element = ReactTestUtils.renderIntoDocument(
      <ClickCounter />,
    );
    const counter = ReactDOM.findDOMNode(element);
    // mouse-down
    ReactTestUtils.Simulate.mouseDown(counter);
    expect(element.state.count).toBe(1);
    expect(element.state.active).toBe(true);
    // mouse-up
    ReactTestUtils.Simulate.mouseUp(counter);
    expect(element.state.count).toBe(1);
    expect(element.state.active).toBe(false);
    // touch-start
    ReactTestUtils.Simulate.touchStart(counter);
    expect(element.state.count).toBe(2);
    expect(element.state.active).toBe(true);
    // touch-end
    ReactTestUtils.Simulate.touchEnd(counter);
    expect(element.state.count).toBe(2);
    expect(element.state.active).toBe(false);
    // press `enter` key
    ReactTestUtils.Simulate.keyDown(counter, {key: "Enter", keyCode: 13, which: 13});
    expect(element.state.count).toBe(3);
    expect(element.state.active).toBe(true);
    // lift key
    ReactTestUtils.Simulate.keyUp(counter);
    expect(element.state.count).toBe(3);
    expect(element.state.active).toBe(false);
  });
});
