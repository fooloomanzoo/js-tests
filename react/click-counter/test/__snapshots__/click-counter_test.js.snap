// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Functionality Renders ClickCounter component correctly 1`] = `
<h1
  aria-label="counter"
  aria-live="polite"
  aria-relevant="text"
  className="basic"
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
>
  Clicks: 
  0
</h1>
`;

exports[`basic renders ClickCounter component correctly 1`] = `
<h1
  aria-label="counter"
  aria-live="polite"
  aria-relevant="text"
  className="basic"
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
>
  Clicks: 
  0
</h1>
`;
